### PR TITLE
Switched APN client for http2 API

### PR DESCRIPTION
**Pull request by sevrugin**
Changed APN client to work with new http2 APN API
Changed other minor issues
Fixed tests

### CODE REVIEW

**Comment by sevrugin** [comment]
@Edujugon is this a chance to have this PR merged?

File: src/Apn.php [hunk 1]
```diff
@@ -6,167 +6,284 @@
...
+            }
+
+            curl_multi_setopt($this->curlMultiHandle, CURLMOPT_PIPELINING, CURLPIPE_MULTIPLEX);
+            if (defined('CURLMOPT_MAX_HOST_CONNECTIONS')) {
+                curl_multi_setopt($this->curlMultiHandle, CURLMOPT_MAX_HOST_CONNECTIONS, $this->maxConcurrentConnections);
+            }
+        }
+
+        $mh = $this->curlMultiHandle;
+        $errors = [];
```
Comment (pwarnes) [inline src/Apn.php:108]: Unused variable

File: src/Apn.php [hunk 2]
```diff
@@ -6,167 +6,284 @@
...
+     * @param array $message
+     * @param $request
+     * @param array $deviceTokens
      */
-    public function getUnregisteredDeviceTokens(array $devices_token)
+    public function prepareHandle($deviceToken, array $message)
     {
-        $tokens = [];
+        $uri = false === $this->config['dry_run'] ? $this->getProductionUrl($deviceToken) : $this->getSandboxUrl($deviceToken);
+        $headers = $message['headers'];
```
Comment (pwarnes) [inline src/Apn.php:255]: potential undefined index

File: src/Apn.php [hunk 3]
```diff
@@ -6,167 +6,284 @@
...
+            CURLOPT_URL => $uri,
+            CURLOPT_PORT => self::APNS_PORT,
+            CURLOPT_POST => true,
+            CURLOPT_POSTFIELDS => $body,
+            CURLOPT_RETURNTRANSFER => true,
+            CURLOPT_TIMEOUT => 10,
+            CURLOPT_HEADER => true,
+
+            CURLOPT_SSLCERT        => $config['certificate'],
+            CURLOPT_SSLCERTPASSWD  => $config['passPhrase'],
```
Comment (pwarnes) [inline src/Apn.php:272]: potential undefined index

**Review by pwarnes — COMMENTED** [review]
Great work! This broke for me a while back and I didn't notice. This PR fixes it (with some tweaks).

The apns-topic is also now a required header so could be good to include support for that so it doesn't need to be set in the message directly.

@Edujugon Please consider merging this when it's ready as your package doesn't work with apns anymore.

**Comment by Edujugon** [comment]
Thanks @sevrugin for this PR 👍 
Great work!!

File: src/Apn.php (hunk 3 above)
Comment (lphhh) [reply src/Apn.php:272]: How do you use it?

Comment (lphhh) [reply src/Apn.php:272]: {\"reason\":\"MissingTopic\"}

Comment (sevrugin) [reply src/Apn.php:272]:
```
        $message = [
            'headers' => [
                'apns-topic' => config('push.apnTopic'),
                'apns-push-type' => 'alert',
            ],
            'aps' => [
                'category' => $pushMessage->getKey(),
            ],
            'data' => $pushMessage->getData(),
        ];
```
like this

Comment (lphhh) [reply src/Apn.php:272]: I found that apnchannel does not have headers. I hope this can be added so that it can be sent through channel

**Pull request merged**